executor/index_lookup_hash_join.go

```diff
@@ -0,0 +1,449 @@
+	// building and its running.
+	// A specific sequence for example:
+	//
+	// e := buildExecutor()   // txn at build time
+	// recordSet := runStmt(e)
+	// session.CommitTxn()    // txn closed
+	// recordSet.Next()
+	// e.dataReaderBuilder.Build() // txn is used again, which is already closed
+	//
+	// The trick here is `getStartTS` will cache start ts in the dataReaderBuilder,
+	// so even txn is destroyed later, the dataReaderBuilder could still use the
+	// cached start ts to construct DAG.
+	_, err := e.innerCtx.readerBuilder.getStartTS()
+	if err != nil {
+		return err
+	}
+
+	err = e.children[0].Open(ctx)
+	if err != nil {
+		return err
+	}
+	e.memTracker = memory.NewTracker(e.id, e.ctx.GetSessionVars().MemQuotaIndexLookupJoin)
+	e.memTracker.AttachTo(e.ctx.GetSessionVars().StmtCtx.MemTracker)
+	e.innerPtrBytes = make([][]byte, 0, 8)
+	e.startWorkers(ctx)
+	return nil
+}
+
+func (e *IndexNestedLoopHashJoin) startWorkers(ctx context.Context) {
+	concurrency := e.ctx.GetSessionVars().IndexLookupJoinConcurrency
+	workerCtx, cancelFunc := context.WithCancel(ctx)
+	e.cancelFunc = cancelFunc
+	innerCh := make(chan *indexHashJoinTask, concurrency)
+	e.workerWg.Add(1)
+	ow := e.newOuterWorker(innerCh)
+	go util.WithRecovery(func() { ow.run(workerCtx) }, e.finishJoinWorkers)
+
+	e.resultCh = make(chan *indexHashJoinResult, concurrency+1)
+	e.joinChkResourceCh = make([]chan *chunk.Chunk, concurrency)
+	for i := int(0); i < concurrency; i++ {
+		e.joinChkResourceCh[i] = make(chan *chunk.Chunk, 1)
+		e.joinChkResourceCh[i] <- newFirstChunk(e)
+	}
+	e.workerWg.Add(concurrency)
+	for i := int(0); i < concurrency; i++ {
+		workerID := i
+		go util.WithRecovery(func() { e.newInnerWorker(innerCh, workerID).run(workerCtx) }, e.finishJoinWorkers)
+	}
+	go e.wait4JoinWorkers()
+}
+
+func (e *IndexNestedLoopHashJoin) finishJoinWorkers(r interface{}) {
+	if r != nil {
+		e.resultCh <- &indexHashJoinResult{err: errors.Errorf("%v", r)}
+	}
+	e.workerWg.Done()
+}
+
+func (e *IndexNestedLoopHashJoin) wait4JoinWorkers() {
+	e.workerWg.Wait()
+	close(e.resultCh)
+}
+
+// Next implements the IndexNestedLoopHashJoin Executor interface.
+func (e *IndexNestedLoopHashJoin) Next(ctx context.Context, req *chunk.Chunk) error {
+	if e.runtimeStats != nil {
+		start := time.Now()
+		defer func() { e.runtimeStats.Record(time.Now().Sub(start), req.NumRows()) }()
+	}
+	req.Reset()
+	result, ok := <-e.resultCh
+	if !ok {
+		return nil
+	}
+	if result.err != nil {
+		return result.err
+	}
+	req.SwapColumns(result.chk)
+	result.src <- result.chk
+	return nil
+}
+
+// Close implements the IndexNestedLoopHashJoin Executor interface.
+func (e *IndexNestedLoopHashJoin) Close() error {
+	if e.cancelFunc != nil {
+		e.cancelFunc()
+		e.cancelFunc = nil
+	}
+	if e.resultCh != nil {
+		for range e.resultCh {
+		}
+		e.resultCh = nil
+	}
+	for i := range e.joinChkResourceCh {
+		close(e.joinChkResourceCh[i])
+		for range e.joinChkResourceCh[i] {
+		}
```

Comment:
It seems no need to wait for `e.joinChkResourceCh[i]` because each worker waiting for this channel will be canceled by `ctx.Done`.

Reply:
The reason is the same as this comment.
https://github.com/pingcap/tidb/pull/8661#discussion_r319792466

Reply:
This line happens after the `cancelFunc` called.